### PR TITLE
WIP: Add static code analysis using PMD

### PR DESCRIPTION
**Pull request by pilino1234**
Very WIP, still working on the rulesets. Comments and feedback very welcome

Add PMD as a Maven plugin, which will run as part of the `verify` stage.
Travis will run PMD to find various issues according to the configured rules. This helps find things such as unused variables and other issues; the [list of available rules](https://pmd.github.io/pmd-6.28.0/pmd_rules_java.html) is long.

### CODE REVIEW

**Comment by pilino1234** [comment]
With just these rules, there are still 76 violations, but this number can easily be raised to 1800 if the codestyle rules that I temporarily excluded are reenabled :smile:

File: travis/pmd-ruleset.xml [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+		<exclude name="MethodArgumentCouldBeFinal"/>
+		<exclude name="OnlyOneReturn"/>
+		<exclude name="ShortClassName"/>
+		<exclude name="ShortVariable"/>
+		<exclude name="TooManyStaticImports"/>
+		<exclude name="UnnecessaryFullyQualifiedName"/>
+		<exclude name="UselessParentheses"/>
+		<exclude name="UseUnderscoresInNumericLiterals"/>
+
+		<!-- TODO: reenable later, but cause a lot of noise -->
```
Comment (pilino1234) [inline travis/pmd-ruleset.xml:35]: I'd like to enable these rules again, but they lead to about 1.7k additional violations currently. Thoughts?

**Comment by pilino1234** [comment]
Adding logs of the violations caused by each set of rules. Hopefully this will make reviewing easier.

Best practices: Coming soon
Codestyle+Documentation: [doc+codestyle-violations.txt](https://github.com/runelite/runelite/files/5321484/doc%2Bcodestyle-violations.txt)
Best practices: Coming soon
Best practices: Coming soon
Best practices: Coming soon

**Comment by pilino1234** [comment]
Adding logs of the violations caused by each set of rules. Hopefully this will make reviewing easier.

Best practices: Coming soon
Codestyle+Documentation: [doc+codestyle-violations.txt](https://github.com/runelite/runelite/files/5321484/doc%2Bcodestyle-violations.txt)
Design: Coming soon
Error prone: Coming soon
Multithreading: Coming soon
Performance: Coming soon
Security: Coming soon

**Comment by pilino1234** [comment]
I added rules from the design section, but I'm not quite sure which ones are actually worth keeping around.
Many the things it complains about would need quite a lot of refactoring work to fix.
See the list of design violations in https://github.com/runelite/runelite/pull/12589#issuecomment-703084799

**Comment by pilino1234** [comment]
I'm not sure if any of the multithreading rules could ever be useful for us. I ended up excluding all the ones that were raising violations because the problems they are warning about aren't particularly applicable to RuneLites threading model anyway (as far as I know, anyway), and the ones that didn't raise violations are even more not useful.
The full list of available multithreading: https://pmd.github.io/pmd-6.28.0/pmd_rules_java_multithreading.html

**Comment by pilino1234** [comment]
Alright, I've gone through and created a preliminary ruleset. It still raises quite a number of violations, but at this point some feedback on which rules to add/remove/reconfigure would be nice, to make this actually useful.

**Comment by Adam-** [comment]
I'd start by getting rid of all of the "too complex" and "too many fields" etc type things. Ideally we are looking only for logic bugs and dead code and maybe resource leaks or threading issues. Most of this still seems subjective.

File: travis/pmd-ruleset.xml (hunk 1 above)
Comment (Nightfirecat) [reply travis/pmd-ruleset.xml:49]: These seem reasonable to me.

**Comment by Nightfirecat** [comment]
I've added another 15 or so exclusions which we don't really care about which drops the violations by another 1k leaving ~700, most of which have pretty quick fixes.

**Comment by pilino1234** [comment]
@Nightfirecat I looked at the violations of the Duplicate Import rule, since there are so few of them I felt that it's probably worth suppressing just those legitimate cases and keeping the rule enabled.

**Comment by pilino1234** [comment]
Finished in 7ceb3c9e020aa8316ddc5809a2d409d5d2efb697.

**Pull request closed without merging**